tmail-backend/tmail-third-party/openpaas/src/main/java/com/linagora/tmail/dav/DavClient.java

```diff
@@ -0,0 +1,307 @@
+ *  https://twake-mail.com/                                         *
+ *  https://linagora.com                                            *
+ *                                                                  *
+ *  This file is subject to The Affero Gnu Public License           *
+ *  version 3.                                                      *
+ *                                                                  *
+ *  https://www.gnu.org/licenses/agpl-3.0.en.html                   *
+ *                                                                  *
+ *  This program is distributed in the hope that it will be         *
+ *  useful, but WITHOUT ANY WARRANTY; without even the implied      *
+ *  warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR         *
+ *  PURPOSE. See the GNU Affero General Public License for          *
+ *  more details.                                                   *
+ ********************************************************************/
+
+package com.linagora.tmail.dav;
+
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.james.util.ReactorUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+import com.linagora.tmail.HttpUtils;
+import com.linagora.tmail.configuration.DavConfiguration;
+import com.linagora.tmail.dav.request.CardDavCreationObjectRequest;
+import com.linagora.tmail.dav.request.GetCalendarByEventIdRequestBody;
+import com.linagora.tmail.dav.xml.DavMultistatus;
+import com.linagora.tmail.dav.xml.DavResponse;
+import com.linagora.tmail.dav.xml.XMLUtil;
+import com.linagora.tmail.james.jmap.model.CalendarEventParsed;
+
+import io.netty.buffer.Unpooled;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.netty.ByteBufMono;
+import reactor.netty.http.client.HttpClient;
+import reactor.netty.http.client.HttpClientResponse;
+import reactor.util.retry.Retry;
+
+public class DavClient {
+    public static final int MAX_CALENDAR_OBJECT_UPDATE_RETRIES = 5;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DavClient.class);
+    private static final Duration DEFAULT_RESPONSE_TIMEOUT = Duration.ofSeconds(10);
+    private static final String COLLECTED_ADDRESS_BOOK_PATH = "/addressbooks/%s/collected/%s.vcf";
+    private static final Pattern CALENDAR_URI_PATTERN = Pattern.compile("/calendars/[^/]+/[^/]+/");
```

Comment:
I got that OpenPaaS calendar URLs are encoded that way.
But I fear that by being too restrictive we actually harm compatibility with 3rd party systems.
Is this validation needed anyway?

Reply:
It is not a process of validation but rather identification. Because the calendars collection could itself contain more collections. This is the least verbose way I have found. An alternative solution is to walk the XML tree and extract the resource type:

```xml
<d:resourcetype>
      <d:collection/>
</d:resourcetype>
```

or 

```xml
 <d:resourcetype>
      <d:collection/>
      <cal:calendar/>
</d:resourcetype>
```